### PR TITLE
add tracing to Mongo access

### PR DESCRIPTION
**Pull request by natebrennand**
This PR adds tracing to our mongo database access.

Upsides:
- easily see how many queries an endpoint makes
- see which of those queries are fast vs slow

Downsides:
- our library interface needs to be updated to encompass every aspect of `mgo.v2` that we utilize
- documentation will make less sense

Next steps:
- roll out more services
- determine any missing tag/log attributes we should add
- _MAYBE_: proactively apply a selector if the destination object has a finite set of fields

### CODE REVIEW

File: mgowrapper.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+		ctx: ts.ctx,
+	}
+}
+
+func (ts tracedMgoSession) Ping() error {
+	sp, _ := opentracing.StartSpanFromContext(ts.ctx, "ping")
+	defer sp.Finish()
+
+	err := ts.sess.Ping()
+	sp.LogFields(opentracinglog.Error(err))
```
Comment (Sayan-) [inline mgowrapper.go]: Looks like we can use `logAndReturnErr` here as well

Comment (natebrennand) [reply mgowrapper.go]: oops, can you tell which one I wrote first? 😜

File: mgohttptest/testing.go [hunk 2]
```diff
@@ -40,8 +41,9 @@ func MakeContext(ctx context.Context, cfgs ...Config) DbHandler {
 	for _, c := range cfgs {
 		newSess := c.Sess.Copy()
 		sessions = append(sessions, newSess)
-		var getSession internal.SessionGetter = func() *mgo.Session {
-			return newSess
+		var getSession internal.SessionGetter = func(ctx context.Context) (*mgo.Session, context.Context) {
+			// add a dummy spam for testing purposes so we don't have a null pointer exception
```
Comment (Sayan-) [inline mgohttptest/testing.go]: `s/spam/span/`

File: mgowrapper.go [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+}
+
+// logAndReturnErr is a tiny helper for adding the error to a log inline.
+func logAndReturnErr(sp opentracing.Span, err error) error {
+	sp.LogFields(opentracinglog.Error(err))
+	return err
+}
+
+// queryToFields transforms an arbitrary mgo arg into a set of log fields.
+// This is mostly geared towards bson.M, but the Sprintf fallback should handle arrays
```
Comment (Sayan-) [inline mgowrapper.go:298]: Should we truncate arrays at some point? Off the top of my head I vaguely remember AVS having some `$in: [giant array]`. I could be wrong

File: session_handler.go [hunk 4]
```diff
@@ -47,6 +51,34 @@ func NewSessionHandler(cfg SessionHandlerConfig) http.Handler {
 	}
 }
 
+// getCallerName is adapted from the runtime pkg examples for the Callers fn
```
Comment (Sayan-) [inline session_handler.go]: might be helpful to have a link for future reference

**Review by Sayan- — COMMENTED** [review]
don't mind me dropping by 👋 

more tracing 🎉 !

File: session_handler.go (hunk 4 above)
Comment (natebrennand) [reply session_handler.go]: https://github.com/Clever/mgohttp/pull/7/commits/81afca6ad3c82a29918f9f03d2bd6751db9b00cf

File: mgohttptest/testing.go (hunk 2 above)
Comment (natebrennand) [reply mgohttptest/testing.go]: https://github.com/Clever/mgohttp/pull/7/commits/81afca6ad3c82a29918f9f03d2bd6751db9b00cf

File: mgowrapper.go (hunk 3 above)
Comment (natebrennand) [reply mgowrapper.go:298]: I think this will be fine, in most cases we limit ourselves to sending no more than 100 ID's at once.

And more generally, this is only inputted as a log which isn't indexed by the tracing provider.... I think
It seems like they only choose to index the "tag" values rather than the "log" values.

File: mgowrapper.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+	return logAndReturnErr(sp, t.db.Run(cmd, result))
+}
+
+type tracedMgoCollection struct {
+	collectionName string
+	collection     *mgo.Collection
+	ctx            context.Context
+}
+
+func (tc tracedMgoCollection) Update(selector interface{}, update interface{}) error {
```
Comment (peternga) [inline mgowrapper.go:66]: What about update multi true?

File: mgowrapper.go [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+	sp, _ := opentracing.StartSpanFromContext(tc.ctx, "update")
+	sp.SetTag("collection", tc.collectionName)
+	sp.LogFields(queryToFields("selector", selector)...)
```
Comment (peternga) [inline mgowrapper.go]: Do we need to worry about sensitive data being logged?

File: mgowrapper.go [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+	info, err = tc.Upsert(selector, update)
+	return info, logAndReturnErr(sp, err)
+}
+
+func (tc tracedMgoCollection) Find(selector interface{}) MongoQuery {
+	sp, ctx := opentracing.StartSpanFromContext(tc.ctx, "find")
+	sp.SetTag("collection", tc.collectionName)
+
+	// NOTE: Find just starts the trace, the finishing call on the MongoQuery must
```
Comment (peternga) [inline mgowrapper.go:113]: Would it make sense to copy+paste this for other calls that just start traces (e.g. Limit)?

File: mgowrapper.go (hunk 5 above)
Comment (natebrennand) [reply mgowrapper.go:66]: Added `UpdateAll` - https://github.com/Clever/mgohttp/pull/7/commits/2615aee3870df473e1b0be59b90c6bc8ba45c1a1

File: mgowrapper.go (hunk 7 above)
Comment (natebrennand) [reply mgowrapper.go:113]: agreed 👍 
added more comments in https://github.com/Clever/mgohttp/pull/7/commits/2615aee3870df473e1b0be59b90c6bc8ba45c1a1

File: mgowrapper.go [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+
+type tracedMgoCollection struct {
+	collectionName string
+	collection     *mgo.Collection
+	ctx            context.Context
+}
+
+func (tc tracedMgoCollection) Update(selector interface{}, update interface{}) error {
+	sp, _ := opentracing.StartSpanFromContext(tc.ctx, "update")
+	sp.SetTag("collection", tc.collectionName)
```
Comment (Sayan-) [inline mgowrapper.go:68]: nit - it'd be nice to use `const` rather than inlines for every tag name

**Review by Sayan- — APPROVED** [review]

File: mgowrapper.go (hunk 6 above)
Comment (natebrennand) [reply mgowrapper.go]: yes, but we no longer will - https://github.com/Clever/mgohttp/pull/7/commits/84e808d1d3a55506bbe3c3bed2b4a0215202866a

**Review by peternga — APPROVED** [review]
